Use doublecross instead of cross(Cross(a, b)) in NavState

scottiyio opened the pull request:
Change cross(cross(a, b)) to doubleCross(a, b) and add documentation

Add a no rotation coriolis test to NavState

scottiyio committed:
Add notes to NavState::coriolis

Change cross(cross(a, b)) to doubleCross(a, b)

dellaert commented on gtsam/navigation/NavState.cpp:
```diff
@@ -444,14 +444,19 @@ Vector9 NavState::coriolis(double dt, const Vector3& omega, bool secondOrder,
   dV(n_xi) << ((-2.0 * dt) * omega_cross_vel);
   if (secondOrder) {
-    const Vector3 omega_cross2_t = omega.cross(omega.cross(n_t));
+    const Vector3 omega_cross2_t = doubleCross(omega, n_t);
```
I think switching to doubleCross is most beneficial if we also grab its Jacobians and use them below to replace some complexity in the Jacobian chaining...

scottiyio replied:
I agree. I did a bit of tests and had one term not matching (probably because I didn't check myself and just trusted copilot) so I will look again tonight 

There's also some doublecrosses elsewhere I believe

scottiyio replied:
Some of this code was rearranged. I wanted to only create doubleCross() once which might be the wrong approach. So H2_wrt_t is created even if we don't use it. Not sure.

scottiyio committed:
Add notes to NavState::coriolis

Change cross(cross(a, b)) to doubleCross(a, b)

scottiyio committed:
- Used doublecross() jacobians
- Remove excess white space

dellaert commented on gtsam/navigation/NavState.cpp:
```diff
@@ -444,23 +444,35 @@ Vector9 NavState::coriolis(double dt, const Vector3& omega, bool secondOrder,
-  // Get perturbations in nav frame
-  Vector9 n_xi, xi;
-  Matrix3 D_dR_R, D_dP_R, D_dV_R, D_body_nav;
+  // Because our navigation frames are placed on a spinning Earth, we experience two apparent forces on our inertials
```
Maybe move this up before sig

dellaert commented on gtsam/navigation/NavState.cpp:
```diff
@@ -444,23 +444,35 @@ Vector9 NavState::coriolis(double dt, const Vector3& omega, bool secondOrder,
   dV(n_xi) << ((-2.0 * dt) * omega_cross_vel);
+
+  // Centrifugal (second order) acceleration corrections
+  Matrix3 H2_wrt_t; // To store Jacobian (if needed/desired)
```
D_c2_nt

dellaert commented on gtsam/navigation/NavState.cpp:
```diff
@@ -444,23 +444,35 @@ Vector9 NavState::coriolis(double dt, const Vector3& omega, bool secondOrder,
-        nRb.unrotate(dV(n_xi), H ? &D_dV_R : 0);
+  // Transform correction from navigation frame -> body frame and get Jacobians
+  Vector9 xi;
+  Matrix3 D_dR_R, D_dP_R, D_dV_R, D_body_nav;
```
D_body_nav -> Rt, and remark that second derivative is the same for all three.

dellaert commented on gtsam/navigation/NavState.cpp:
```diff
@@ -444,23 +444,35 @@ Vector9 NavState::coriolis(double dt, const Vector3& omega, bool secondOrder,
+  dP(xi) = nRb.unrotate(dP(n_xi), H ? &D_dP_R : 0);
+  dV(xi) = nRb.unrotate(dV(n_xi), H ? &D_dV_R : 0);
 
+  // Assemble Jacobians
```
Should be D_p_v etc..... we use p here not t

dellaert commented on gtsam/navigation/NavState.cpp:
```diff
@@ -472,9 +484,8 @@ Vector9 NavState::coriolis(double dt, const Vector3& omega, bool secondOrder,
-      const Matrix3 D_cross2_state = Omega * D_cross_state;
-      D_t_t(H) -= D_body_nav * (0.5 * dt2) * D_cross2_state;
-      D_v_t(H) -= D_body_nav * dt * D_cross2_state;
+      D_t_t(H) -= (0.5 * dt2) * D_body_nav * (H2_wrt_t * R());
```
D_p_p(H) -= (0.5 * dt2) * D_c2_np; ???

dellaert commented on gtsam/navigation/NavState.cpp:
```diff
@@ -472,9 +484,8 @@ Vector9 NavState::coriolis(double dt, const Vector3& omega, bool secondOrder,
-      D_t_t(H) -= D_body_nav * (0.5 * dt2) * D_cross2_state;
-      D_v_t(H) -= D_body_nav * dt * D_cross2_state;
+      D_t_t(H) -= (0.5 * dt2) * D_body_nav * (H2_wrt_t * R());
+      D_v_t(H) -= dt * D_body_nav * (H2_wrt_t * R());
```
dt * D_c2_np; ????

I'd add a unit test with non-identity R before - make sure it fails - then apply fix, should succeed.

dellaert commented:
Would love to merge this!

scottiyio committed:
Simplifed code by removing R() from the code. Resolved some extra declarations.

Rotated omega (earth angular rate, not IMU angular velocities) to the body frame for simplification

Cannot rename "D_t_t" to say "D_p_p" or "D_v_t" to "D_v_p" because of NavState derivatives?
